File: music-application/src/main/java/com/music/sale/application/product/mapper/ProductMapper.java
package com.music.sale.application.product.mapper;

import com.music.sale.application.product.dto.input.CreateProductInput;
import com.music.sale.application.product.dto.input.UpdateProductImageInput;
import com.music.sale.application.product.dto.input.UpdateProductInput;
import com.music.sale.application.product.dto.output.ProductImageOutput;
import com.music.sale.application.product.dto.output.ProductOutput;
import com.music.sale.domain.product.ProductImage;
import com.music.sale.domain.product.ProductItem;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class ProductMapper {
  public ProductItem toDomainForCreate(CreateProductInput input, Long createdBy) {
    return ProductItem.builder()
        .catalogId(input.catalogId())
        .sellerId(input.sellerId())
        .storeId(input.storeId())
        .name(input.name())
        .brand(input.brand())
        .price(input.price())
        .condition(input.condition())
        .conditionGrade(input.conditionGrade())
        .stockQuantity(input.stockQuantity())
        .status(input.status())
        .attributes(input.attributes())
        .description(input.description())
        .viewCount(0L)
        .build();
  }

  public ProductItem toDomainForUpdate(ProductItem existing, UpdateProductInput input) {
    return existing.toBuilder()
        .name(getValueOrDefault(input.name(), existing.getName()))
        .brand(getValueOrDefault(input.brand(), existing.getBrand()))
        .catalogId(getValueOrDefault(input.catalogId(), existing.getCatalogId()))
        .storeId(getValueOrDefault(input.storeId(), existing.getStoreId()))
        .price(getValueOrDefault(input.price(), existing.getPrice()))
        .condition(getValueOrDefault(input.condition(), existing.getCondition()))
        .conditionGrade(getValueOrDefault(input.conditionGrade(), existing.getConditionGrade()))
        .stockQuantity(getValueOrDefault(input.stockQuantity(), existing.getStockQuantity()))
        .status(getValueOrDefault(input.status(), existing.getStatus()))
        .attributes(getValueOrDefault(input.attributes(), existing.getAttributes()))
        .description(getValueOrDefault(input.description(), existing.getDescription()))
        .build();
  }

  private <T> T getValueOrDefault(T value, T defaultValue) {
    return value != null ? value : defaultValue;
  }

  public ProductOutput toOutput(ProductItem item) {
    return new ProductOutput(
        item.getId(),
        item.getCatalogId(),
        item.getSellerId(),
        item.getStoreId(),
        item.getName(),
        item.getBrand(),
        item.getPrice(),
        item.getCondition(),
        item.getConditionGrade(),
        item.getStockQuantity(),
        item.getStatus(),
        item.getAttributes(),
        item.getDescription(),
        item.getViewCount());
  }

  public ProductImageOutput toProductImageOutput(ProductImage image) {
    return new ProductImageOutput(
        image.getId(),
        image.getProductId(),
        image.getUrl(),
        image.isThumbnail(),
        image.getImageOrder(),
        image.getFileSize(),
        image.getFileName(),
        image.getFileType());
  }

  public ProductImageOutput toProductImageOutput(
      ProductImage savedImage, UpdateProductImageInput input) {
    return new ProductImageOutput(
        savedImage.getId(),
        savedImage.getProductId(),
        savedImage.getUrl(),
        savedImage.isThumbnail(),
        savedImage.getImageOrder(),
        input.fileSize(),
        input.fileName(),
        input.fileType());
  }
}

File: music-application/src/main/kotlin/com/music/sale/application/product/mapper/ProductMapper.kt
// Copyright (C) 2024 Your Name or Company
package com.music.sale.application.product.mapper

import com.music.sale.application.product.dto.*
import com.music.sale.domain.product.Product
import com.music.sale.domain.store.Store
import com.music.sale.domain.user.User
import org.springframework.stereotype.Component

@Component
class ProductMapper {
    fun toOutput(product: Product): ProductOutput {
        return ProductOutput(
            id = product.id,
            catalog =
                ProductCatalog(
                    id = product.catalog.id,
                    name = product.catalog.name,
                    categories = listOf(product.catalog.category),
                    brand = product.catalog.brand ?: "",
                    attribute = product.catalog.attributes?.mapValues { it.value.toString() } ?: emptyMap(),
                ),
            name = product.name(),
            price = product.price,
            seller = product.seller,
            store = product.store,
            condition = product.condition,
            conditionGrade = product.conditionGrade,
            stockQuantity = product.stockQuantity,
            status = product.status,
            attributes = product.attributes(),
            images = product.images.toList(),
        )
    }

    fun toSearchProductCondition(input: SearchProductInput): SearchProductCondition {
        return SearchProductCondition(
            keyword = input.keyword,
            keywordType = input.keywordType,
            categoryId = input.categoryId,
            location = input.location,
            condition = input.condition,
            conditionGrade = input.conditionGrade,
            minPrice = input.minPrice,
            maxPrice = input.maxPrice,
            status = input.status,
        )
    }

    fun toSaveProductCondition(
        input: CreateProductInput,
        seller: User?,
        store: Store?,
    ): SaveProductItemCondition {
        return SaveProductItemCondition(
            name = input.name,
            catalogId = input.catalogId,
            price = input.price,
            seller = seller,
            store = store,
            condition = input.condition,
            conditionGrade = input.conditionGrade,
            stockQuantity = input.stockQuantity,
            status = input.status,
            attributes = input.attributes,
        )
    }
}
